### DISCUSSION TITLE
Ignore interface in flows for virtual interfaces

### DISCUSSION THREAD

**Discussion by rpravm** [akvorado/akvorado#386] [Q&A]
I look like that Akvorado always checks in/out interface's oid in flows.

But I have an interesting case - I'm running netflow on Juniper router with BNG enabled, and I have traffic coming to subscribers with their own unique interfaces, they have oids like 200076315.
SNMP queries for this interfaces are disabled and will be a problem at scale of tens thousands.

Maybe there can be an option for disabling strict checking for both interfaces in flow? Or maybe another workaround?

**Comment by vincentbernat** [comment]
Without interface information, you cannot classify the interface and therefore, you cannot know the direction. How would you know in your case if a flow is ingress or egress?

**Reply by rpravm** [reply]
Maybe it will be meaningful to save a flow with only one interface?

**Reply by vincentbernat** [reply]
That's already the case. Only one interface is required.

**Reply by rpravm** [reply]
Ok, I see some flows with one interface empty, thanks.
I made some checks with Wireshark and found that this works only if interface value in flow = 0.

But in this my case flows have both fields: valid InIf but OutIf with unknown oid - my router ignores SNMP requests with it.
So they are ignored by inlet.

Can this be fixed anyhow?

<img width="400" alt="image" src="https://user-images.githubusercontent.com/35486468/211204983-6234a5b3-1a6e-4af1-a013-b2b48f58ea48.png">

**Reply by vincentbernat** [reply]
Oh, you are right, it works when one interface is missing, but not when an interface contains something that's not solvable by SNMP (mostly because I don't want to have bad flows when the system is starting). Also, there is no negative cache for SNMP, so you will get a lot of SNMP traffic because of the missing interface information.

A long term solution would be to make the SNMP part pluggable. One can choose to use SNMP, HTTP, or static files to solve names and descriptions. This would be helpful for other people too.

In the meantime, if you can compile your own version (it's just "docker build"), you can just modify `enricher.go` with a condition like: 

```go
if flow.InIf > 2000000000 {
  flow.InIfName = "something"
  flow.InIfDescription = "something"
  flow.InIfSpeed = uint32(100000)
} else if flow.InIf != 0 {
  ...
}
```
You still need one valid interface to get exporter name.

**Reply by vincentbernat** [reply]
Would you be able to generate a JSON file with the mappings? Or is it also something not possible (like ifIndex is not predictible or too volatile).

Also, Akvorado has a cache and only queries a few values. It's not like a classic NMS that will poll counters and stuff like that. Maybe you can provide a specific community where SNMP is enabled for these interfaces.

**Reply by rpravm** [reply]
Thanks you a lot for the suggestions, I think that condition is a good solution for my case, I'll try to use it. 
Using pre-defined iface maps or snmp would not be nice in my case.

Thanks for great product by the way! :)

**Reply by vincentbernat** [reply]
What would be the ideal solution for your case, assuming anything is possible?

**Reply by rpravm** [reply]
I solved this by modifying source - called this interfaces by special name and description.

Making some configuration option for this tweak - looks like it would be counterintuitive and my case is very rare.

1. Maybe you can add some checking of interfaces - for example if flow has one interface detected in SNMP and other does not - sign this interface as "Unknown".
2. Or maybe there is any sense in making option for "Templated" interface, for example if oid > 200000000 then iface will be named "Templated". Or some kind of regex search by oid?
3. Make option "Allow one valid interface" and second will be made "Templated" from config

All these ways are not very good I think.. :))
But anyway - skipping that flows is a bad thing too. I spent some days finding why I don't some see traffic :()

**Reply by vincentbernat** [reply]
An alternative would be to have an option allowing unknown interfaces (or maybe an interface with a negative cache entry) and expose the index to the classifier for people to fill the blanks. This is partly related to #384.

**Comment by vincentbernat** [comment]
This should be available in next version. SNMP not knowing an interface is not fatal anymore and you can classify using interface index as well. See #488.
